sshd-core/src/main/java/org/apache/sshd/client/session/ClientConnectionService.java

```diff
@@ -55,7 +61,41 @@ public ClientConnectionService(AbstractClientSession s) throws SshException {
 
         heartbeatRequest = CoreModuleProperties.HEARTBEAT_REQUEST.getRequired(this);
         heartbeatInterval = CoreModuleProperties.HEARTBEAT_INTERVAL.getRequired(this);
-        heartbeatReplyMaxWait = CoreModuleProperties.HEARTBEAT_REPLY_WAIT.getRequired(this);
+        heartbeatMaxNoReply = configureMaxNoReply();
+    }
+
+    protected int configureMaxNoReply() {
+        @SuppressWarnings("deprecation")
+        Duration timeout = CoreModuleProperties.HEARTBEAT_REPLY_WAIT.getOrNull(this);
+        if (timeout == null || GenericUtils.isNegativeOrNull(heartbeatInterval) || GenericUtils.isEmpty(heartbeatRequest)) {
+            return CoreModuleProperties.HEARTBEAT_NO_REPLY_MAX.getRequired(this).intValue();
+        }
+        // The deprecated timeout is configured explicitly. If the new no-reply-max is _not_ explicitly configured,
+        // set it from the timeout.
+        Integer noReplyValue = CoreModuleProperties.HEARTBEAT_NO_REPLY_MAX.getOrNull(this);
+        if (noReplyValue != null) {
+            return noReplyValue.intValue();
+        }
+        if (timeout.compareTo(heartbeatInterval) >= 0) {
+            // Timeout is longer than the interval. With the previous system, that would have killed the session when
+            // the
```

Comment:
This should be reformatted to avoid a single word on a line.

Reply:
Done.